### PR TITLE
Verify that the key matches the cert

### PR DESCRIPTION
**Pull request by aimuz**
Signed-off-by: aimuz <mr.imuz@gmail.com>

#### What type of PR is this?

/kind bug
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Verify that the key matches the cert
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
This change can result in existing secrets being treated as invalid and failing updates. I think the most we could do here is emit a warning.

**Comment by aimuz** [comment]
> This change can result in existing secrets being treated as invalid and failing updates. I think the most we could do here is emit a warning.

Thanks, this will result in a non-backward-compatible situation.

I've reworked it to pass the warning

File: pkg/registry/core/secret/strategy.go [hunk 1]
```diff
@@ -130,3 +133,15 @@ func SelectableFields(obj *api.Secret) fields.Set {
 	}
 	return generic.MergeFieldsSets(objectMetaFieldsSet, secretSpecificFieldsSet)
 }
+
+func warningsForSecret(secret *api.Secret) []string {
+	var warnings []string
+	if secret.Type == api.SecretTypeTLS {
+		// Verify that the key matches the cert.
+		_, err := tls.X509KeyPair(secret.Data[api.TLSCertKey], secret.Data[api.TLSPrivateKeyKey])
```
Comment (liggitt) [inline pkg/registry/core/secret/strategy.go:141]: I'm trying to think if we construct key pairs from API-provided info elsewhere, and whether passing arbitrary user-provided data into X509KeyPair exposes more of an attack surface than we have previously. @tallclair, do you happen to know?

Comment (tallclair) [reply pkg/registry/core/secret/strategy.go:141]: My feeling on this is that the permission to write secrets is high enough that we shouldn't be designing around the risk of CVEs in the go standard library code. This isn't performing a security-critical function, so the worst case would be a DoS vulnerability (which has maybe happened before with CVE-2022-24675), but I still think that given the `editor` level role required it's an acceptable risk (lots of ways to DoS a cluster with the editor role...)

**Pull request merged**